### DISCUSSION TITLE
PTScotch needs flex installed error while installing Cardinal

### DISCUSSION THREAD

**Discussion by nancysemwal** [idaholab/moose#29017] [Q&A Getting Started] [answered]
### Check these boxes if you have followed the posting rules.

- [X] I have consulted the [Posting Guidelines](https://github.com/idaholab/moose/discussions/18270).
- [X] I have searched the [Discussions Forum](https://github.com/idaholab/moose/discussions) and [MOOSE Framework Troubleshooting](https://mooseframework.inl.gov/help/troubleshooting.html) and have not found what I was looking for
- [X] Q&A Getting Started is the most appropriate category for my question (trouble installing, beginner user, ...)

### Issue or question about MOOSE

Hello everyone.

I am installing Cardinal on my PC using [this](https://cardinal.cels.anl.gov/without_conda.html) "Building Cardinal without MOOSE's conda Environment" guide. 

Before installing Cardinal, I had already installed MOOSE using `conda`  because I didn't know that Cardinal could manage its dependencies itself. I can access this MOOSE installation using `conda activate moose`. However, while installing Cardinal, I am not in this environment. I am in the base environment only. So I think I'm on the right path.

However when executing `./contrib/moose/scripts/update_and_rebuild_petsc.sh` command I am getting the following error:            

`
UNABLE to CONFIGURE with GIVEN OPTIONS (see configure.log for details):- PTScotch needs flex installed
`

I have already searched this forum and found [Questions when building Cardinal without MOOSE's conda environment #27115](https://github.com/idaholab/moose/discussions/27115) which talked about deinit and rinit using `git submodule update --init --recursive moose` command (This didn't work for me) and [Error when installing PETSc and libMesh #27184](https://github.com/idaholab/moose/discussions/27184) but here the issue was with `METIS` and not `flex` so I didn't try the solutions. So I'm making a new thread.

One thing to note is that I am not doing the installation in my `HOME` directory but rather in some other directory. I believe that's not a problem as long as I specify the correct paths.

Changes made in my `.bashrc` - 

`export NEKRS_HOME=/mnt/faa6b7ba-bc59-4346-b975-2b5cd78afa12/cardinal/install`

`export CC=mpicc`
`export CXX=mpicxx`
`export FC=mpif90`

Here I have omitted one command which is suggested as OPTIONAL in the tutorial 
`export OPENMC_CROSS_SECTIONS=${HOME}/cross_sections/endfb-vii.1-hdf5/cross_sections.xml`

That's because I am not installing in `HOME`, and when I checked the path where I am installing (`/mnt/faa6b7ba-bc59-4346-b975-2b5cd78afa12/`), there is no folder named `cross_sections`. Could this be the cause of my issue?

I also checked my `configure.log` file and saw something which seems amiss to me. 

`Configure Options: --with-hdf5-dir=/home/nancy/.local/share/Trash/files/anaconda2/pkgs/hdf5-1.10.2-hba1933b_1/include/hdf5.h/../../ `

Why is the `HDF5` package being picked up from a trash directory ? I am assuming this is somehow responsible for the issue.

I feel like I have tried everything that I could. What more can be done?

Details:
OS - Ubuntu 24.04.1 LTS (Installed about 2 months ago)
Cmake - 3.28.4
GCC - 13.2.0
G++ - 13.2.0
GFortran - 13.2.0
Git - 2.43.0
Conda - 24.9.1

### Encountering Errors? Please include diagnostic output

Influential Environment Variables

`CC=mpicc`
`CONDA_CHANNEL=https://conda.software.inl.gov/public`
`CONDA_DEFAULT_ENV=base`
`CONDA_EXE=/home/nancy/miniconda3/bin/conda`
`CONDA_PREFIX=/home/nancy/miniconda3`
`CONDA_PROMPT_MODIFIER=(base)`
`CONDA_PYTHON_EXE=/home/nancy/miniconda3/bin/python`
`CONDA_SHLVL=1`
`CURL_CA_BUNDLE=`
`CXX=mpicxx`
`FC=mpif90`
`LD_LIBRARY_PATH=/usr/local/cuda-11.8/lib64
`MOOSE_JOBS=6`
`PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin`
`REQUESTS_CA_BUNDLE=`
`SSL_CERT_FILE=`
`Compiler(s) (CC CXX FC F77 F90):`

`which $CC; /usr/bin/mpicc`
`$CC --version; gcc (Ubuntu 13.2.0-23ubuntu4) 13.2.0`
`$CC -show; gcc -I/usr/lib/x86_64-linux-gnu/openmpi/include -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -L/usr/lib/x86_64-linux-gnu/openmpi/lib -lmpi`

`which $CXX; /usr/bin/mpicxx`
`$CXX --version; g++ (Ubuntu 13.2.0-23ubuntu4) 13.2.0`
`$CXX -show; g++ -I/usr/lib/x86_64-linux-gnu/openmpi/include -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -L/usr/lib/x86_64-linux-gnu/openmpi/lib -lmpi_cxx -lmpi`

`which $FC; /usr/bin/mpif90`
`$FC --version; GNU Fortran (Ubuntu 13.2.0-23ubuntu4) 13.2.0`
`$FC -show; gfortran -I/usr/lib/x86_64-linux-gnu/openmpi/lib/../../fortran/gfortran-mod-15/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/lib -L/usr/lib/x86_64-linux-gnu/openmpi/lib/fortran/gfortran -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi -lopen-rte -lopen-pal -lhwloc -levent_core -levent_pthreads -lm -lz`

`F77	not set`
`F90	not set`

`FAIL: One or more compiler environment variables not set, or set incorrectly`

`Python Sanity Checks`

`/usr/bin/env python3 --version; (reporting as: Python 3.12.3) matches`
`which python3 python;`

`/usr/bin/python3 --version; == Python 3.12.3`
`WARNING: python; does not exist`

OK

Python Modules (TestHarness, run-ability)

`FAIL:    missing module(s): packaging `

`Either install the above packages, or perhaps you have yet
to activate the moose environment: conda activate moose`

`Missing 'failing' Python modules will prevent you from building MOOSE.`

`FAIL: One or more Python issues present.`

`MOOSE Repository/Conda Version Checks`

`WARNING: Unable to run Conda tests due to missing Python modules`

`checks FAILED`

**Comment by GiudGiud** [comment, marked as answer]
The path forward here is to install flex.
It should just be `sudo apt-get install flex`

> there is no folder named cross_sections. Could this be the cause of my issue?

No this wont cause issues for installation. You ll need cross sections to run openmc later though

> Why is the HDF5 package being picked up from a trash directory 

Do you have it installed elsewhere?

**Reply by milljm** [reply]
It is because we add an inclusive path to `~/.local`. ref: https://github.com/idaholab/moose/blob/facbaefe7429e714a38735bf85a6d2ed0958698a/scripts/configure_petsc.sh#L52

@cticenhour and I spoke about possibly refining the path to be not so broad.

**Reply by nancysemwal** [reply]
I installed `flex` using `sudo apt-get install flex` and also deleted the trash folder from where the `HDF5` package was getting picked. After this, there was no problem in building the 3 dependencies. But I am now stuck in a new problem which again involves `HDF5`, detailed [here](https://github.com/idaholab/moose/discussions/29043).
